pkg/collector/corechecks/servicediscovery/discover_services_linux.go

```diff
@@ -0,0 +1,318 @@
+	}
+
+	log.Infof("alive: %d | services: %d | ignore: %d", len(processes), len(i.alive.m), len(i.ignore.m))
+	aliveProcesses := make(map[int]struct{})
+
+	var wg sync.WaitGroup
+
+	for _, p := range processes {
+		aliveProcesses[p.PID] = struct{}{}
+		name, shortName, err := processName(p)
+		if err != nil {
+			log.Errorf("[pid: %d] failed to get process name: %v", p.PID, err)
+			continue
+		}
+
+		ignore := false
+		if ignoreName, ok := i.ignore.get(p.PID); ok {
+			// ignore
+			if ignoreName == name {
+				ignore = true
+			} else {
+				// same pid but different process name, this means the pid has been reused for
+				// a different process
+				i.ignore.delete(p.PID)
+			}
+		} else if _, ok := i.ignoreCfg[shortName]; ok {
+			ignore = true
+			i.ignore.set(p.PID, name)
+		}
+
+		if !ignore {
+			// process is a potential service
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				i.scanProcess(p.PID, openPorts)
+			}()
+		}
+	}
+
+	for pid, pInfo := range i.alive.m {
+		_, ok := aliveProcesses[pid]
+		if !ok {
+			// TODO: if there is a start and a stop for the same process name in this iteration, it means the process
+			// 	has been restarted. In that case we should only send one of the events.
+			i.sender.sendEndServiceEvent(pInfo)
+			i.alive.delete(pid)
+		}
+	}
+
+	// cleanup ignore processes that are no longer alive
+	for pid, _ := range i.ignore.m {
+		_, ok := aliveProcesses[pid]
+		if !ok {
+			i.ignore.delete(pid)
+		}
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (i *linuxImpl) scanProcess(pid int, openPorts portlist.List) {
+	p, err := procfs.NewProc(pid)
+	if err != nil {
+		log.Errorf("failed to get proc: %v", err)
+		return
+	}
+
+	if pInfo, ok := i.alive.get(pid); ok {
+		if time.Since(pInfo.LastHeatBeatTime) >= heartbeatTime {
+			i.sender.sendHeartbeatServiceEvent(pInfo)
+			pInfo.LastHeatBeatTime = time.Now()
+		}
+		return
+	}
+
+	pInfo, err := getProcessInfo(p, openPorts)
+	if err != nil {
+		log.Errorf("failed to get process info: %v", err)
+		return
+	}
+
+	if len(pInfo.Ports) == 0 {
+		log.Infof("[pid: %d | name: %s]: process has no open ports, ignoring...", pid, pInfo.ShortName)
+		i.ignore.set(pid, pInfo.Name)
+		return
+	}
+
+	dCtx := servicedetector.New(pInfo.CmdLine, pInfo.Env)
+	meta, ok := dCtx.Detect()
+	if !ok {
+		log.Warnf("failed to detect service (pid: %d, name: %s)", pInfo.PID, pInfo.ShortName)
+		return
+	}
+
+	if len(meta.AdditionalNames) > 0 {
+		for _, n := range meta.AdditionalNames {
+			pInfo.Services = append(pInfo.Services, &serviceInfo{Name: n})
+		}
+	} else {
+		pInfo.Services = append(pInfo.Services, &serviceInfo{Name: meta.Name})
+	}
+
+	i.alive.set(pid, pInfo)
+	i.sender.sendStartServiceEvent(pInfo)
+}
+
+/*
+- command line /proc/{pid}/cmdline
+- environment variables /proc/{pid}/environ
+- CWD (PWD env var)
+- open ports /proc/{pid}/net/tcp|udp|tcp6|udp6
+- time launched /proc/{pid}/stat
+*/
+func getProcessInfo(p procfs.Proc, openPorts portlist.List) (*processInfo, error) {
```

Comment:
I think we discussed a different algorithm that should be simpler.

- get map of ports that are listened on to associated PIDs
   - for each PID
      - if we haven’t seen it before, put on “potential” list
      - if it is on the “potential” list, add to “service” list, remove from “potential” list, send start-service event
      - if it is on the “service” list and it’s been 15 minutes, send heartbeat-service event
  - for each PID on the “service” list that wasn’t in the map of ports->PIDs
     - send end-service event
     - remove from “service” list
- sleep for 1 minute